include/wil/safearrays.h

```diff
@@ -0,0 +1,1105 @@
+    //!     RETURN_IF_FAILED(data.access(psa));
+    //!     result.reserve(result.size() + data.size());
+    //!     for (BSTR& bstr : data)
+    //!     {
+    //!         result.emplace_back(wil::make_bstr_nothrow(bstr));
+    //!         RETURN_HR_IF_NULL(E_OUTOFMEMORY, result.rbegin()->get());
+    //!     }
+    //!     return S_OK;
+    //! }
+    //! ~~~~
+    //! wil::unique_bstr_safearray move_to_safearray(std::vector<wil::unique_bstr>& source)
+    //! {
+    //!     auto sa = wil::unique_bstr_safearray{source.size()};
+    //!     auto current = std::begin(source);
+    //!     for (BSTR& bstr : sa.access_data())
+    //!     {
+    //!         // Transfer ownership of the BSTR into the safearray
+    //!         bstr = (*current++).release();
+    //!     }
+    //!     return sa;
+    //! }
+    //! ~~~~
+    //! void handle_safearray_of_bools(SAFEARRAY* psa, std::function<void(bool, ULONG)> fnHandler)
+    //! {
+    //!     auto data = wil::safearraydata<VARIANT_BOOL>{psa};
+    //!     for (auto i = 0U; i < data.size(); ++i)
+    //!     {
+    //!         // Translate to Regular Bool and Invoke the Handler
+    //!         fnHandler(data[i] != VARIANT_FALSE, i);
+    //!     }
+    //! }
+    //! ~~~~
+    //! template<typename T>
+    //! std::vector<T> extract_all_values(SAFEARRAY* psa)
+    //! {
+    //!     auto data = wil::safearraydata<T>{psa};
+    //!     // Construct the vector from the iterators
+    //!     return {data.begin(), data.end()};
+    //! }
+    //! ~~~~
+    //! template<typename INTERFACE>
+    //! std::vector<wil::com_ptr_t<INTERFACE, wil::err_exception_policy>> extract_all_interfaces(SAFEARRAY* psa)
+    //! {
+    //!     auto result = std::vector<wil::com_ptr_t<INTERFACE, wil::err_exception_policy>>{};
+    //!     auto data = wil::safearraydata<LPUNKNOWN>{psa};
+    //!     result.reserve(data.size());
+    //!     for (auto& p : data) // Type of p is LPUNKNOWN
+    //!     {
+    //!         // Use "tag_com_query" if you want failure instead of nullptr if INTERFACE not supported
+    //!         result.emplace_back(p, wil::details::tag_try_com_query);
+    //!     }
+    //!     return result;
+    //! }
+    //! ~~~~
+    //! @tparam T           The underlying datatype expected in the SAFEARRAY. It should be a scalar type, BSTR, VARIANT_BOOL, LPUNKNOWN/LPDISPATCH, etc.
+    //! @tparam err_policy  Represents the error policy for the class (error codes, exceptions, or fail fast; see @ref page_errors)
+    template <typename T, typename err_policy = err_exception_policy>
+    class safearraydata_t
+    {
+    public:
+        typedef typename err_policy::result result;
+        typedef safearray_unaccess_scope_exit::pointer pointer;
+        typedef T value_type;
+        typedef value_type* value_pointer;
+        typedef const value_type* const_value_pointer;
+        typedef value_type& value_reference;
+        typedef const value_type& const_value_reference;

```

Comment:
```suggestion
        typedef T value_type;
        typedef value_type* pointer;
        typedef const value_type* const_pointer;
        typedef value_type& reference;
        typedef const value_type& const_reference;
```
Am I missing something? Is there a reason to deviate from the typical naming conventions used in the STL and other libraries? You can replace the uses of `pointer` below with just `SAFEARRAY*` - you shouldn't need a typedef for it (nor do you really need to reference its type in such an obscure way via `safearray_unaccess_scope_exit::pointer`)